# Table of Contents for 4.7 Deriving New Variables
    Import libraries, set path, import files (ords_prods_merge.pkl)
    Lesson: Check shape
            Create a subset and check the shape
            Make a definition for the range of prices as price_label
            Apply it to the subset as new column price_range and check its counts
            Check the max price on it
            If-Statements with the loc() Function:
                Using loc() to make a new column price_range_loc
                    Locate  the column prices and for any values over 15 assign High-range product 
                    Less than or equal to 15 but less than 5 assign Mid-range product
                    Less than or equal to 5 assign Low-range product
                Get value_counts on price_range_loc
                Use the loc() definitions to make a new column price_range_loc on the full dataset and get the value_counts
                Check the columns of the full df
                Check the value_counts (frequency) for orders_day_of_week
            If-Statements with For-Loops:
                Create a list that will loop through orders_day_of_week and return a string
                Use the list now as a column 'busiest_day' in the dataframe
                Check busiest_day value_counts
            Cross check both new columns: orders_day_of_week and busiest_day (shapes)
    Task:   Create a list 'results' to loop through orders-of-the-week and return strings. 
                Busiest_day is 0. Next_busiest_day is 1. Slowest_day is 4 and Next_slowest_day is 3
                Use the list on df as busiest_days
                Check value_counts and shape
            Identify the busiest hours of the day
            Create a list 'busy_hour' to loop through orders-of-the-week and return strings.
                Use the list on df as busiest_day_of_week
                Get shape and frequency
            Review number of columns
            Export as ord_prods_merge_newvariable.pkl      
            

# Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Set path

In [2]:
path = r'C:\Users\valev\Desktop\0325 Instacart Basket Analysis'

# Import files

In [3]:
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ord_prods_merge.pkl' ))


In [4]:
df_ords_prods_merge.shape

(32404859, 14)

# Create a subset

In [5]:
df=df_ords_prods_merge[:1000000]

In [6]:
df.shape

(1000000, 14)

# Define price_label to then flag a column labeled price_range

In [7]:
# Define the function
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [8]:
# 'Apply the function
df['price_range']=df.apply(price_label, axis=1)

C:\Users\valev\AppData\Local\Temp\ipykernel_21304\2266790169.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range']=df.apply(price_label, axis=1)


In [9]:
# Check counts for each return in price_range
df['price_range'].value_counts(dropna=False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

In [10]:
#Check max price in subset since there are supposedly zero counts for high range but my results do show them
df['prices'].max()

24.5

## Why do I get a high range count and my max is 24.5

# If-Statements with the loc() Function

In [11]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


C:\Users\valev\AppData\Local\Temp\ipykernel_21304\2857404532.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [12]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 


In [13]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [14]:
df['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     652638
Low-range product     338018
High-range product      9344
Name: count, dtype: int64

# Use loc() on entire dataframe not just subset

In [15]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [16]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [17]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [18]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

### Why do I get different counts!!!!

# Frequency of orders on each day of the week

In [19]:
df_ords_prods_merge.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'price_range_loc'],
      dtype='object')

In [20]:
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

# If-Statements with For-Loops

## Create a list that will loop through orders_day_of_week and return a string

In [21]:

result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [22]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Regularly busy

## Use the list now as a column 'busiest_day' in the dataframe

In [23]:
df_ords_prods_merge['busiest_day']=result

In [24]:
df_ords_prods_merge['busiest_day'].value_counts(dropna=False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

## Cross check both new columns

In [25]:
df_ords_prods_merge['orders_day_of_week'].shape

(32404859,)

In [26]:
df_ords_prods_merge['busiest_day'].shape

(32404859,)

In [27]:
df_ords_prods_merge.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'price_range_loc', 'busiest_day'],
      dtype='object')

In [28]:
df_ords_prods_merge.shape

(32404859, 16)

# TASK

## Create a list 'results' to loop through orders-of-the-week and return strings.  Busiest_day is 0. Next_busiest_day is 1.  Slowest_day is 4 and Next_slowest_day is 3.  

In [29]:
results = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    results.append("Busiest day")
  elif value == 4:
    results.append("Least busy")
  elif value == 1:
     results.append("Next_busiest_day")
  elif value == 3:
     results.append("Next_slowest_day")
  else:
    results.append("Regularly busy")

In [30]:
results

['Next_busiest_day',
 'Next_busiest_day',
 'Busiest day',
 'Regularly busy',
 'Next_busiest_day',
 'Regularly busy',
 'Next_busiest_day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Next_slowest_day',
 'Regularly busy',
 'Regularly busy',
 'Next_slowest_day',
 'Regularly busy',
 'Regularly busy',
 'Next_slowest_day',
 'Regularly busy',
 'Next_slowest_day',
 'Next_busiest_day',
 'Least busy',
 'Next_busiest_day',
 'Regularly busy',
 'Next_busiest_day',
 'Regularly busy',
 'Regularly busy',
 'Next_busiest_day',
 'Least busy',
 'Next_slowest_day',
 'Next_busiest_day',
 'Busiest day',
 'Next_busiest_day',
 'Next_busiest_day',
 'Least busy',
 'Least busy',
 'Next_busiest_day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Next_slowest_day',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Next_slowest_day',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest

# Use the results list on a new column in dataframe as busiest_days -- later I can drop busiest_day if not needed by customer

In [31]:
df_ords_prods_merge['busiest_days']=results

In [32]:
df_ords_prods_merge['busiest_days'].value_counts(dropna=False)

busiest_days
Regularly busy      12916111
Busiest day          6204182
Next_busiest_day     5660230
Next_slowest_day     3840534
Least busy           3783802
Name: count, dtype: int64

In [33]:
df_ords_prods_merge.shape

(32404859, 17)

In [34]:
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

## Now the customer can see the definition on a new column as they wished.   And the values match as expected. 

#  Identify the busiest hours of the day

In [35]:
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

# Create a list 'busy_hour' to loop through orders-of-the-week and return strings. 
### Most_orders will include hours 10 down to 12  - from list above
### Average_orders will include from hours 16 down to 19
### Fewest_orders will include from hourse 20 down to 3

In [36]:
busy_hour = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10,11,14,15,13,12]:
    busy_hour.append("Most orders")
  elif value in [19,18,8,17,9,16]:
    busy_hour.append("Average orders")
  elif value in[3,4,2,5,1,0,6,23,22,21,7,20]:
    busy_hour.append("Fewest orders")

  else:
    busy_hour.append("No information")
      

## Use the list on df and get frequency on the new column

In [37]:
df_ords_prods_merge['busiest_period_of_day'] = busy_hour

In [38]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna=False)

busiest_period_of_day
Most orders       16128666
Average orders    11689984
Fewest orders      4586209
Name: count, dtype: int64

In [39]:
df_ords_prods_merge.shape

(32404859, 18)

# The dataframe now has 18 columns

In [40]:
df_ords_prods_merge.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'price_range_loc', 'busiest_day', 'busiest_days',
       'busiest_period_of_day'],
      dtype='object')

## It may be possible to drop _merge, busiest_day but will wait for future exercises

# Export to pkl

In [41]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ord_prods_merge_newvariable.pkl'))